### PR TITLE
feat(caching): add SnapshotStorage

### PR DESCRIPTION
**Pull request by dlovell**

### CODE REVIEW

**Review by mesejo — APPROVED** [review]

File: python/letsql/common/caching.py [hunk 1]
```diff
@@ -155,3 +163,33 @@ def _put(self, key, value):
...
+                tokenized = dask.base.tokenize(expr)
+                return "-".join(("snapshot", tokenized))
+
+    @staticmethod
+    def normalize_backend(con):
+        name = con.name
+        if name in ("pandas", "duckdb", "datafusion"):
+            return (name, None)
+        else:
+            return normalize_backend(con)
```
Comment (mesejo) [inline python/letsql/common/caching.py:185]: If the backend name is `postgres,` wouldn't this method reach `RecursionError`?

File: python/letsql/common/utils/dask_normalize.py [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
 
+
 dask.config.set({"tokenize.ensure-deterministic": True})
+
+
+@contextmanager
+def patch_normalize_token(*typs, f=toolz.functoolz.return_none):
+    with patch.dict(
+        dask.base.normalize_token._lookup,
+        {typ: Mock(side_effect=f) for typ in typs},
```
Comment (mesejo) [inline python/letsql/common/utils/dask_normalize.py:22]: Out of curiosity, why is the code using testing mocks?

**Review by mesejo — APPROVED** [review]

File: python/letsql/common/caching.py (hunk 1 above)
Comment (dlovell) [reply python/letsql/common/caching.py:185]: the `normalize_backend` it accesses is the one imported, `self.normalize_backend` **would** result in a `RecursionError`

File: python/letsql/common/utils/dask_normalize.py (hunk 2 above)
Comment (dlovell) [reply python/letsql/common/utils/dask_normalize.py:22]: for `SnapshotStorage` to work, we need to remove some specific sources of non-determinism in `dask.base.normalize_token` but only during the context of specific invocations. `patch` fits this use and we already need this capability for tests.

**Pull request merged**